### ISSUE TITLE
Spotify player stop working

### ISSUE THREAD

**Issue by henriquetorres** [berrberr/streamkeys#99]
Since a few days ago, this stop working. I's using default key bindings.

osX 10.10.3 with chrome v 43.0.2357.81 (64-bit)

**Comment by berrberr** [comment]
Yeah, a couple of people have been reporting this but I can't reproduce - it is still working for me. URL I'm using is: https://play.spotify.com/

Couple of questions:
1. Is the URL you are using https://play.spotify.com/ ?
2. Are you a paid or free user?
3. Can you upload a screenshot of what your player looks like? The one I'm seeing is: http://i.imgur.com/xPgPoBg.jpg

My guess right now is that they are doing rolling tests of a new player (and maybe they target paid users first?)

**Comment by henriquetorres** [comment]
Hi,

1 - Yes, thats the one I'm using, but it is redirected to: https://player.spotify.com/browse
2 - Free user
3 -  Here it is: http://imgur.com/QjU6LzO

**Comment by berrberr** [comment]
Thanks for the info! It does look like that player is indeed different. I can't figure out how to get to it though ...  https://player.spotify.com/browse redirects me back to https://play.spotify.com/browse

I think that the player you are seeing is the beta player or something...and my account isn't in it

**Comment by henriquetorres** [comment]
Hi,

Yah, probably :)

If it helps, here it is the html generated for controls on this "new/beta" player:

<code>
<div role="complementary" data-bind="attr: {'aria-label': PLAYER_LABEL}" class="player-controls-container flex-row-container" aria-label="Player controls">
      <a id="player-anchor" class="anchor" tabindex="-1"></a>
        <div id="controls">
          <button type="button" id="previous" data-non-active-element="" class="button button-icon-only button-no-border spoticon-skip-back-32" data-bind="
            click: skipToPrevious,
            attr: {
              disabled: !previousEnabled(),
              'data-tooltip': previousLabel,
              'aria-label': previousLabel
            }
            " data-tooltip="Previous" aria-label="Previous"></button>
          <button type="button" id="play" data-non-active-element="" class="button button-icon-only button-play playing" data-bind="css: {
              playing: playing
            },
            click: togglePlay,
            attr: {
              disabled: !playEnabled(),
              'data-tooltip': playLabel,
              'aria-label': playLabel
            }
            " data-tooltip="Pause" aria-label="Pause"></button>
          <button type="button" id="next" data-non-active-element="" class="button button-icon-only button-no-border spoticon-skip-forward-32" data-bind="attr: {
              disabled: !nextEnabled(),
              'data-tooltip': nextLabel,
              'aria-label': nextLabel
            },
            click: skipToNext" data-tooltip="Next" aria-label="Next"></button>
        </div>
        <span id="elapsed" data-bind="text: elapsed">2:50</span>
        <div id="progressbar" class="flex-1"><div class="progress-bar-wrapper handle">
  <div class="progress-bar">
    <div class="inner" style="right: 8.37914967028361%;">
      <div class="handle"></div>
    </div>
  </div>
</div>
</div>

        <span id="remaining" data-bind="text: remaining, click: toggleRemainingTimeFormat">3:06</span>
        <span id="lyrics-button" data-bind="text: LYRICS_LABEL, click: openLyrics">Lyrics</span>
        <div id="extra-buttons">
          <button type="button" id="queue" name="queue" class="button button-icon-only button-no-border spoticon-queue-16" data-bind="
            attr: {
              'data-drop-target': queueDropMimetypes,
              'data-tooltip': queueLabel,
              'aria-label': queueLabel
            },
            click: openQueue
            " data-drop-target="text/x-spotify-tracks|text/x-spotify-albums|text/x-spotify-artists|text/x-spotify-playlists" data-tooltip="Queue" aria-label="Queue"></button>
          <button type="button" id="shuffle" name="shuffle" data-non-active-element="" class="button button-icon-only button-no-border spoticon-shuffle-16" data-bind="
              css: {
                active: shuffle,
                disabled: !shuffleEnabled()
              },
              click: toggleShuffle,
              attr: {
                value: shuffle,
                'data-tooltip': shuffleLabel,
                'aria-label': shuffleLabel
              }" data-tooltip="Shuffle" aria-label="Shuffle"></button>
          <button type="button" id="repeat" name="repeat" data-non-active-element="" class="button button-icon-only button-no-border spoticon-repeat-16" data-bind="
              css: repeatIcon() + (repeatEnabled() ? '' : ' disabled'),
              click: toggleRepeat,
              attr: {
                value: repeat,
                'data-tooltip': repeatLabel,
                'aria-label': repeatLabel
              }" data-tooltip="Repeat" aria-label="Repeat" value="0"></button>
          <button type="button" id="connecticon" name="connect" class="button button-icon-only button-no-border spoticon-volume-16 button-no-hover" data-bind="
              css: volumeIcon(),
              click: onClickAtVolumeButton,
              attr: {
                'data-tooltip': connectLabelOrEmpty,
                'aria-label': connectLabelOrEmpty
              },
              event: {
                mousewheel: volumebar.onMouseWheel
              }" data-tooltip="" aria-label=""></button>
          <div id="connect-picker" class="dropdown" tabindex="-1" data-bind="
                with: connectPopup,
                css: { visible: connectPopup.visible }
              ">
            <div id="connect-container" class="dropdown-menu arrow-bottom arrow-position-right" data-bind="css: {enabled: $root.connectEnabled}">
              <div id="connect-title" data-bind="visible: devicesVisible" style="display: none;">
                <span data-bind="text: title">CHOOSE DEVICE TO PLAY ON</span>
                <img class="help" data-bind="click: showHelp, attr: {'data-tooltip': helpHintText}" src="images/connect_help.png" data-tooltip="What is Connect?">
              </div>
              <div id="connect-help" data-bind="visible: helpVisible" style="display: none;">
                <div>
                  <img class="help" data-bind="click: showHelp, attr: {'data-tooltip': helpHintText}" src="images/connect_help.png" data-tooltip="What is Connect?">
                  <img class="speakericon" src="images/connect_logo.png">
                  <p data-bind="text: helpText1">Connect lets you play and control Spotify on your devices.</p>
                  <p data-bind="text: helpText2">Start Spotify on another device and it will magically appear here.</p>
                </div>
              </div>
              <div id="connect-offline" data-bind="visible: offlineVisible" style="display: none;">
                <div>
                  <img class="help" data-bind="click: showHelp, attr: {'data-tooltip': helpHintText}" src="images/connect_help.png" data-tooltip="What is Connect?">
                  <img src="images/connect_offline.png" width="81" height="48">
                   <span class="title" data-bind="text: connectOfflineTitle">You're offline!</span>
                   <p data-bind="text: helpText1">Connect lets you play and control Spotify on your devices.</p>
                   <p data-bind="text: connectOfflineText1">You need to go online in order to use this feature.</p>
                </div>
              </div>
              <div id="connect-upsell" data-bind="visible: upsellVisible" style="display: none;">
                <div>
                  <img class="speakericon" src="images/connect_logo.png">
                  <p data-bind="text: upsellText">Connect is a Premium feature that lets you play and control Spotify on your devices.</p>
                </div>
                <button class="button" data-bind="click: onShowMore, text: showMoreLabel">Show More</button>
                <button class="button button-primary" data-bind="click: onUpgrade, text: upgradeLabel">Upgrade</button>
              </div>
              <div id="connect-devices" data-bind="visible: devicesVisible" style="display: none;">
                <ul data-bind="foreach: devices"></ul>
              </div>
            </div>
          </div>
        </div>
        <div id="volumebar-container" data-bind="event: {
          mousewheel: volumebar.onMouseWheel
        }">
          <div id="volumebar">
            <div class="progress-bar-wrapper handle" data-bind="
                visible: volumebar.visible,
                css: {
                  handle: volumebar.enabled,
                  active: volumebar.thumb.active
                },
                event: {
                  mousedown: volumebar.onMouseDown
                }">
              <div class="progress-bar">
                <div class="inner" data-bind="style: {right: volumebar.remaining}" style="right: 0%;">
                  <div class="handle"></div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div id="offline">
          <span data-bind="text:offlineLabel">Offline</span>
          <i class="spoticon-offline-16" data-bind="attr: {'data-tooltip': offlineLabel}" data-tooltip="Offline"></i>
        </div>
      </div>
</code>

**Comment by matteosecli** [comment]
Hi, I'm experiencing the same problem. It works fine on https://play.spotify.com but not on https://player.spotify.com.
Just a note about the redirection. If I go on player.spotify.com, sometimes it redirects me on play.spotify.com; however, editing the url and forcing "player.spotify.com" again does the trick. Also, sometimes just reloading the page on play.spotify.com redirects to player.spotify.com.
Anyway, maybe you can get the new web player code via the .deb package for Linux; the interface is the same!
I'm on Ubuntu 14.04, if you need additional info or testing just let me know!
Very nice and complete extension, by the way! :)

**Comment by berrberr** [comment]
Yeah, so it is a targeted release that users can't control (see [the Spotify forums](https://community.spotify.com/t5/Help-Desktop-Linux-Mac-Windows/How-do-i-switch-the-webplayer-UI-back-to-the-older-one/m-p/1143520#M129688)). So I'll be unable to do any development until I get a hold of a "beta" account, which could be a while...

@matteosecli @henriquetorres : if either of you are interested in developing support for the new player let me know! I very recently wrote some code to handle multiple different players on the same site (see: https://github.com/berrberr/streamkeys/blob/master/code/js/controllers/YoutubeController.js).

It should be really easy to add, all you need to do is find the new selectors. The "magic" happens in the `checkPlayer` function which uses a query selector to see which player is active. I could scaffold the code in `SpotifyController` if you are able to find the new selectors & test it. LMK

**Comment by henriquetorres** [comment]
Hi berrberr, isn't the new selectors on the code I've pasted here http://pastebin.com/cV4JiBqP ?

At this moment, I can't access https://player.spotify.com it just redirects me back to https://play.spotify.com

**Comment by berrberr** [comment]
@henriquetorres Whoops I overlooked that, I think that that will be enough to build something for you to test

**Comment by berrberr** [comment]
Hey @henriquetorres @matteosecli , I just pushed an update to master to try to get it to work with the new UI. If either of you could test it that would be great :+1: (I still don't have access to new player!)

**Comment by henriquetorres** [comment]
Hello @berrberr, at the moment I have SK v1.4.8, is this the version with the new implementation? If yes, it's not working :(

**Comment by matteosecli** [comment]
Hi @berrberr, I'm sorry for the late response but I've been busy with some university exams. :)
I'll test your update asap and let you know!

**Comment by matteosecli** [comment]
Hi again @berrberr. I've built the updated code but it seems that it's not working. Yet it's strange, because the selectors look correct. I've taken a screenshot:
![schermata del 2015-06-15 16 48 11](https://cloud.githubusercontent.com/assets/4618521/8162618/c357ce48-137e-11e5-9d11-f2c26d3c2b1c.png)
As you see, it seems to work in the '#main' iframe. It's the same if I just select the element with the dev tools:
![schermata del 2015-06-15 16 40 23](https://cloud.githubusercontent.com/assets/4618521/8162680/28de679a-137f-11e5-9817-b2326f011592.png)
I've also tried to modify the Spotify controller to use only the new selectors, following your previous version of [SpotifyController.js](https://github.com/berrberr/streamkeys/blob/f48608f2d11d50c8d50f9bb005e2deb7a614293d/code/js/controllers/SpotifyController.js) as a template, but it doesn't work anyway.
I'm not quite an expert with these things, so maybe it's just me! :)

**Comment by berrberr** [comment]
@henriquetorres No you will have to manually add it to chrome. Here is a link to the built extension: https://dl.dropboxusercontent.com/u/8406200/streamkeys149.zip . [Here](http://superuser.com/questions/247651/how-does-one-install-an-extension-for-chrome-browser-from-the-local-file-system) is a tutorial on how to load a chrome extension from folder.

@matteosecli No problem :) Can you open up the console and see if any errors are being logged when you try all of the player actions? Can you also verify that the new player controls are inside an iframe with CSS selector `#main`?

Thanks for the help all!

**Comment by matteosecli** [comment]
@berrberr I think that the new player controls are in the #main iframe; If you look at the first picture I've sent in the previous post, the frame "main (index.html)" is selected in the console. I've also tried to open a new tab at the address contained in the #main description, and I've tried again with a CSS selector; the controls seem to be there! :)
Anyway I think there is something weird; I've enabled the opts logging (I hope I've done it right!) and then I took a look at the console; the extension seems to be loading a controller for Player.fm and so the selectors are all wrong (and there is no iframe). I've taken a screenshot:
![schermata del 2015-06-15 19 10 22](https://cloud.githubusercontent.com/assets/4618521/8165731/1c0191aa-1393-11e5-90b5-48001023e361.png)
I've also saved the log in a file: http://pastebin.com/B9g1mSsz

**Comment by berrberr** [comment]
Ohhhh I see what's happening now, great find @matteosecli . It's actually injecting the controller for `player.fm` since the url is `player.spotify.com` (and unless it is overridden with a more specific regex it defaults to just searching the hostname).

Try this: change this line https://github.com/berrberr/streamkeys/blob/master/code/js/modules/Sitelist.js#L58 to:
`"player.fm": {name: "Player.fm", url: "http://player.fm", controller: "PlayerController.js", enabled: true},` & rebuild.

This should stop the wrong controller from being injected and let us actually test the correct one!

**Comment by henriquetorres** [comment]
Hi @berrberr I've just checked and I've the same result as @matteosecli reported.
I tried to solve it but could figured it out, I've already tried to a more specific  regex on Sitelist.js:72, with player.fm but same problem, on file background.js:53 "get_site_controller" I've forced that function to return a valid response for spotify (response('SpotifyController.js');) and noticed that the new controller was assumed but the was an error on querySelector, after some debuf I've noticed you have on SpotifyController.js:36 doc.querySelector and it must be document.querySelector. And controlls are working again :dancers:  :D

**Comment by matteosecli** [comment]
I can confirm what @henriquetorres find out: correcting line 58 in Sitelist.js as suggested by @berrberr and changing line 35 in SpotifyController.js to "document.querySelector" does the trick and the controls are working again! :D
Anyway, it gives an error that prevents it from compiling:
```
code/js/controllers/SpotifyController.js
  line 33  col 9  'doc' is defined but never used.

✖ 1 problem

Warning: Task "jshint:all" failed. Use --force to continue.

Aborted due to warnings.
```
It compiles using --force, but still gives the warning.

**Comment by henriquetorres** [comment]
@matteosecli I think you just need to change checkPlayer to this:

  controller.checkPlayer = function() {
    var that = this,
        doc = document.querySelector(multiSelectors.iframe[0]) || document.querySelector(multiSelectors.iframe[1]);

    if(doc.querySelector(multiSelectors.playPause[0])) {
      $.each(multiSelectors, function(key, value) {
        that.selectors[key] = value[0];
      });
    }
    else {
      $.each(multiSelectors, function(key, value) {
        that.selectors[key] = value[1];
      });
    }
  };
})();


But in my case changing line 58 to player.fm dind't solve problem, I've only testes the new spotify controller with success because I've forced to always return SpotifyController.js on background.js, just to see if the new binds where working

**Comment by matteosecli** [comment]
@henriquetorres it compiles & works correctly with your suggestion. :)

Have you also added ```controller: "PlayerController.js"``` to line 58 of Sitelist.js? It's weird, because it works in my case with just that correction.

**Comment by henriquetorres** [comment]
Humm,

It wasn't working because @berrberr dropbox version was defining sitelist on background.js so my change were always overriden :)

Now it is working perfectly :+1:

**Comment by berrberr** [comment]
Yay! Thanks @matteosecli  @henriquetorres for all the help :smile:.

I'll be pushing these fixes to the Chrome webstore tonight in v1.4.9!
